ttnn/cpp/ttnn/operations/experimental/ccl/all_gather_matmul/device/multi_core/all_gather_matmul_op_multi_core.cpp

```diff
@@ -0,0 +1,210 @@
+// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+///
+#include <algorithm>
+
+#include "tt_metal/common/core_coord.h"
+#include "eth_l1_address_map.h"
+#include "impl/buffers/buffer.hpp"
+#include "ttnn/tensor/tensor_impl.hpp"
+#include "ttnn/operations/ccl/all_gather/device/all_gather_op.hpp"
+#include "ttnn/operations/ccl/shared_with_host/hetergeneous_data_structs.hpp"
+#include "ttnn/operations/ccl/ccl_host_datastructures.hpp"
+#include "ttnn/operations/ccl/ccl_common.hpp"
+#include "ttnn/deprecated/tt_dnn/op_library/math.hpp"
+#include "ttnn/deprecated/tt_dnn/op_library/work_split.hpp"
+#include "tt_metal/common/constants.hpp"
+#include "tt_metal/detail/util.hpp"
+#include "tt_metal/host_api.hpp"
+#include <sstream>
+#include <type_traits>
+
+#include "ttnn/cpp/ttnn/operations/experimental/ccl/all_gather_matmul/device/all_gather_matmul_op.hpp"
+#include "ttnn/operations/experimental/ccl/ccl_op_fusion.hpp"
+
+
+using namespace tt::constants;
+
+namespace ttnn {
+
+using namespace experimental::ccl;
+using Tensors = std::vector<Tensor>;
+
+std::tuple<std::vector<CoreCoord>, std::vector<uint32_t>, std::optional<operation::OverrideRuntimeArgumentsCallback<Tensors>>> setup_datacopy(
```

Comment:
I'd suggest to use struct, and create typename aliases. Looks too large to understand.